Adds Toggle, Variables, User Link and Delete to Inventory Host/Host Details

AlexSCorey opened the pull request:
##### SUMMARY
This addresses issue #5266 

The user can navigate from host list or inventory Host list to host details.  They can toggle details on/off and they can return to the list.  It also adds links to the user that created/modified the host details.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

##### AWX VERSION
```
9.1.0
```
##### ADDITIONAL INFORMATION
 If the user comes to Host details through Inventory Host they will get a
Return To Host tab in addition to the others.  This PR allows Inventory Host
to share many of the same components with Host but does add some complexity
to the routing files in Host.jsx.  This does not address Inventory Host Edit which I think @marshmalien  is working on.  It also does not address https://github.com/ansible/awx/issues/5599


![HostDetails](https://user-images.githubusercontent.com/39280967/71839640-a976a800-3089-11ea-9ff0-443af853eb1b.gif)

AlexSCorey committed:
Adds Toggle, Variables, user Link and Delete to Inventory Host and Host Details

 If the user comes to Host details through Inventory Host they will get a
Return To Host tab in addition to the others.  This PR allows Inventory Host
to share many of the same components with Host but does add some complexity
to the routing files in Host.jsx

AlexSCorey commented on awx/ui_next/src/screens/Host/Host.jsx:
```diff
@@ -23,8 +26,16 @@ class Host extends Component {
+    this.handleHostToggle = this.handleHostToggle.bind(this);
+    this.handleToggleError = this.handleToggleError.bind(this);
+    this.handleHostDelete = this.handleHostDelete.bind(this);
+    this.toggleDeleteModal = this.toggleDeleteModal.bind(this);
```
I chose to keep all of the delete modal toggling and delete functionality since this component already had state.  I was also trying to keep HostDetails stateless

marshmalien replied:
In my opinion, these methods belong within the context of HostDetail.jsx. Host.jsx acts primarily as a "route" file for the subtab views. Since only the HostDetail view cares about these methods, it would make more sense for the delete and toggle logic to live in there. This would also simplify the Host.jsx file and reduce the number of props to drill down.

jlmitch5 replied:
Just want to make sure...does this follow the UX of other error/delete on detail views.  what @marshmalien is making sense of where it should live, but I also want to make sure we aren't doing something "special" here that we're not doing in other places that have the same concepts.

AlexSCorey replied:
I was able to move most of this work into `HostDetail` file which reduced the props being passed in. I believe that `Host` is now ready to be refactored into a functional component but that can be done during another PR.

AlexSCorey replied:
@jlmitch5 
> Just want to make sure...does this follow the UX of other error/delete on detail views. what @marshmalien is making sense of where it should live, but I also want to make sure we aren't doing something "special" here that we're not doing in other places that have the same concepts.

JobDetail (handles deletion errors as well as delete), InventoryDetail (handles delete but does not handle deletion errors), InventoryGroupDetail (handle delete and deletion errors)

OrganizationDetail, UserDetail JobTemplateDetail have no delete functionality at all, and probably should follow the same pattern as the files mentioned above.  

So for Detail components to have delete functionality those components handle the delete on their own.  I think our pattern is still a bit shaky but should be that detail components handle their own delete and their own deletion errors.

AlexSCorey commented on awx/ui_next/src/screens/Host/Host.jsx:
```diff
@@ -45,31 +56,99 @@ class Host extends Component {
+      if (history.location.pathname.startsWith('/hosts')) {
+        setBreadcrumb(data);
+      }
+      setBreadcrumb(inventory, data);
```
I had to write this because if the user is coming from Inventory Hosts the breadcrumb needs to know the inventory info as well as the host info to render properly.

marshmalien replied:
Since this line follows the **if** block, it will always get called regardless of the location pathname. This means that if the user is at `/#/hosts/1/details`, the code will enter the **if** block, call the first `setBreadcrumb`, then call the second `setBreadcrumb`. Since `inventory` is undefined at `/#/hosts/1/details`, when `setBreadcrumb` gets called in Hosts.jsx, the `host` parameter will be undefined.

AlexSCorey commented on awx/ui_next/src/screens/Host/Host.jsx:
```diff
@@ -45,31 +56,99 @@ class Host extends Component {
-      { name: i18n._(t`Groups`), link: `${match.url}/groups`, id: 2 },
+      {
+        name: i18n._(t`Return to Hosts`),
+        link: `/inventories/inventory/${match.params.id}/hosts`,
```
come back to this

AlexSCorey commented on awx/ui_next/src/screens/Host/Host.jsx:
```diff
@@ -117,62 +196,99 @@ class Host extends Component {
-            )}
-            {host && (
+      <>
+        <PageSection
```
Things got complicated with the PageSections in Inventory and Host.  This css prop eliminates extra padding if the user is coming from Inventories

AlexSCorey commented on awx/ui_next/src/screens/Host/Host.jsx:
```diff
@@ -117,62 +196,99 @@ class Host extends Component {
-            {host && (
+      <>
+        <PageSection
+          css={`
```
Things got complicated with the PageSections in Inventory and Host.  This css prop eliminates extra padding if the user is coming from Inventories

marshmalien replied:
Alex and I chatted offline about how to remove this css prop by recomposing the components.

AlexSCorey commented on awx/ui_next/src/screens/Host/HostDetail/HostDetail.jsx:
```diff
@@ -42,20 +148,12 @@ function HostDetail({ host, i18n }) {
-        />
+        <Detail label={i18n._(t`Created`)} value={createdBy} />
+        <Detail label={i18n._(t`Last Modified`)} value={modifiedBy} />
         <VariablesDetail
```
There is an issue with this here https://github.com/ansible/awx/issues/5599

AlexSCorey commented on awx/ui_next/src/screens/Inventory/Inventory.jsx:
```diff
@@ -61,10 +62,14 @@ function Inventory({ history, i18n, location, match, setBreadcrumb }) {
     location.pathname.endsWith('edit') ||
     location.pathname.endsWith('add') ||
-    location.pathname.includes('groups/')
+    location.pathname.includes('groups/') ||
```
For this line and the line below the cardHeader is generated elsewhere.  This is necessary because these are nested resources which require different card header/tabs than what is generated here.  We have to generate those different tabs elsewhere because we don't have required information in this component.

AlexSCorey commented on awx/ui_next/src/screens/Inventory/InventoryHosts/InventoryHostList.jsx:
```diff
@@ -0,0 +1,226 @@
+import React, { useEffect, useState } from 'react';
```
This is just a file rename.  I believe it was called InventoryHosts, or Hosts. No code has changed here.

AlexSCorey commented on awx/ui_next/src/components/RoutedTabs/RoutedTabs.jsx:
```diff
@@ -57,23 +57,25 @@ function RoutedTabs(props) {
-          }
-        />
-      ))}
+      {tabsArray
```
This block below allows us to add a 'Return to ____' tab and render it in in certain nested contexts like Inventory Groups or Inventory Hosts

marshmalien replied:
We _could_ move that logic into the component that builds the tab array instead of adding a condition in routed tabs. That would keep the routed tabs component more generic and remove the need for `isNestedTab`.

For instance, we could define the basic "Host" tabsArray: [ Details, Facts, Groups, Completed Jobs] in Host.js. If the location doesn't start with `/hosts`, then we could modify the tabsArray to show: [ Back to Hosts, Details, Facts, Groups, Completed Jobs].

jlmitch5 replied:
agreed with @marshmalien.  If we _can_ keep the custom stuff in the calling component, I think that's preferred for now.

I talked with taufique about this one, and he mentioned the "return to..." pattern will only be on specifically nested places like hosts, and not a pattern throughout the app, so I don't think we need to involve the generic component for the handful of instances we will have to do this on.

AlexSCorey commented on awx/ui_next/src/screens/Host/Host.jsx:
```diff
@@ -45,31 +56,99 @@ class Host extends Component {
+        name: i18n._(t`Return to Hosts`),
+        link: `/inventories/inventory/${match.params.id}/hosts`,
+        id: 99,
+        isNestedTab: !history.location.pathname.startsWith('/hosts'),
```
This is used in Routed Tabs to know when to filter this tab.

marshmalien replied:
Shouldn't `Back to Hosts` always be a nested tab?

marshmalien replied:
Shouldn't `Back to Hosts` always be a nested tab?

AlexSCorey replied:
@marshmalien 
Yes it should always be a nested tab.  Here I am trying to give RoutedTabs something to key off of to know whether to show that tab.  Since this tabs array is shared by `/hosts/:id` and `/inventories/inventory/:id/hosts/:hostId` I needed to be able to hide that "Back to Hosts" tab when we are in `/hosts/:id`

AlexSCorey commented on awx/ui_next/src/screens/Inventory/Inventories.jsx:
```diff
@@ -27,7 +27,7 @@ class Inventories extends Component {
   }
 
-  setBreadCrumbConfig = (inventory, group) => {
+  setBreadCrumbConfig = (inventory, nestedResource) => {
```
We are passing this function down several levels at this point.  We should probably consider thinking about refactoring this somehow.

marshmalien replied:
Yeah, we should probably look into rewriting this in the future. We could leverage the use of hooks in Breadcrumbs 2.0. 

The breadcrumbConfig below has a few lines of duplication. To clean it up, I would group the related breadcrumbs, arrange by order of appearance, and remove any duplication.

marshmalien commented on awx/ui_next/src/screens/Host/HostDetail/HostDetail.jsx:
```diff
@@ -64,11 +162,25 @@ function HostDetail({ host, i18n }) {
           )}
+        {summary_fields.user_capabilities &&
+          summary_fields.user_capabilities.delete && (
+            <Button
```
We could probably use the `< DeleteButton />` component here. I saw an example of it in the InventoryDetail view.

marshmalien commented on awx/ui_next/src/screens/Host/Host.jsx:
```diff
@@ -45,31 +56,99 @@ class Host extends Component {
-      { name: i18n._(t`Facts`), link: `${match.url}/facts`, id: 1 },
-      { name: i18n._(t`Groups`), link: `${match.url}/groups`, id: 2 },
+      {
+        name: i18n._(t`Return to Hosts`),
```
This should say ["Back to Hosts"](https://tower-mockups.testing.ansible.com/patternfly/inventories/inventories-hosts-detail/)

marshmalien commented on awx/ui_next/src/screens/Host/Hosts.jsx:
```diff
@@ -64,6 +67,7 @@ class Hosts extends Component {
                     location={location}
                     setBreadcrumb={this.setBreadcrumbConfig}
                     me={me || {}}
+                    inventory={inventory}
```
When would `inventory` have a value here?

AlexSCorey replied:
It doesn't. Removed

AlexSCorey replied:
It doesn't.  Removed

AlexSCorey committed:
Improves NestedTabs, Refactors PR, Adds Delete/DeleteError Functionality to HostDetail

marshmalien commented on awx/ui_next/src/screens/Host/HostDetail/HostDetail.jsx:
```diff
@@ -18,13 +23,114 @@ const ActionButtonWrapper = styled.div`
+  let createdBy = '';
+  if (created) {
+    if (summary_fields.created_by && summary_fields.created_by.username) {
+      createdBy = (
```
We could use the `< UserDateDetail />` component for the "created and modified by" details.

marshmalien commented on awx/ui_next/src/screens/Host/HostDetail/HostDetail.test.jsx:
```diff
@@ -50,8 +50,7 @@ describe('<HostDetail />', () => {
     const wrapper = mountWithContexts(<HostDetail host={mockHost} />);
-    // VariablesDetail has two buttons
-    const editButton = wrapper.find('Button').at(2);
+    const editButton = wrapper.find('Button[aria-label="edit"]');
```
🎉 A more solid test selector!

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryHosts/InventoryHosts.jsx:
```diff
@@ -1,227 +1,46 @@
-  const isAllSelected = selected.length > 0 && selected.length === hosts.length;
+import Host from '../../Host/Host';
+import InventoryHostList from './InventoryHostList';
+import HostAdd from '../InventoryHostAdd';
```
Since we already have a component named `HostAdd`, I don't think we should rename this import from `InventoryHostAdd` to `HostAdd`.

jlmitch5 commented on awx/ui_next/src/screens/Host/Host.jsx:
```diff
@@ -117,62 +196,99 @@ class Host extends Component {
+                  ]}
+                  render={() => (
+                    <HostDetail
+                      match={match}
```
do we pass match in other places?  I feel like the better pattern would be to wrap the HostDetail.jsx component in the react-router provider, but I'm not sure if this is what we do everywhere, and that's why this is done this way.

This does feel like a lot of props...that suggestion of moving that stuff to the HostDetail function would make this simpler.

AlexSCorey replied:
@jlmitch5 
no, we don't pass match in other places I have already removed that, and have been able to reduce the number of props going to Host Details to 2(host, and onUpdateHost, which is a function that HostDetails uses to update host.)

jlmitch5 commented on awx/ui_next/src/screens/Host/Hosts.jsx:
```diff
@@ -46,14 +46,17 @@ class Hosts extends Component {
-          <Route path={`${match.path}/add`} render={() => <HostAdd />} />
+          <Route
+            path={`${match.path}/add`}
+            render={() => <HostAdd history={history} />}
```
again, I'd use the provider, not pass via props.

jlmitch5 reviewed (approved):
This is looking good, and I think some of the improvements Marly is suggesting will make it even better.  Great job on this so far!

AlexSCorey commented:
recheck

AlexSCorey committed:
Addresses Console Errors related to functions and test data

Also Adds speecificity to link URLs by add /details for urls that should
redireect to details pages instead of them ending in /:id

AlexSCorey commented:
recheck

marshmalien commented on awx/ui_next/src/components/AddRole/SelectResourceStep.jsx:
```diff
@@ -109,6 +109,7 @@ class SelectResourceStep extends React.Component {
                   name={item[displayKey]}
                   label={item[displayKey]}
                   onSelect={() => onRowClick(item)}
+                  onDeselect={() => {}}
```
Please change this from a noop function to `onRowClick()`.

marshmalien commented on awx/ui_next/src/screens/Host/HostDetail/HostDetail.jsx:
```diff
@@ -68,4 +157,4 @@ HostDetail.propTypes = {
 };
 
-export default withI18n()(HostDetail);
+export default withI18n()(withRouter(HostDetail));
```
Since we are trying to replace the withRouter HOC with react-router hooks in our functional components, please remove the HOC and retrieve the params and history object from the `useParams` and `useHistory` hooks.

AlexSCorey committed:
Utilizes React Router Hooks and removes No-op function

marshmalien commented on awx/ui_next/src/screens/Inventory/Inventories.jsx:
```diff
@@ -39,33 +39,46 @@ class Inventories extends Component {
       ),
       [`/inventories/${inventoryKind}/${inventory.id}/hosts`]: i18n._(t`Hosts`),
-      [`/inventories/${inventoryKind}/${inventory.id}/hosts/add`]: i18n._(
-        t`Create New Host`
```
We need to add this route back "Create New Host"

AlexSCorey committed:
Fixes Breaedcrumb

dsesami reviewed (approved):
Tried out the host toggle in the details page. sparkline, responsiveness, etc. work. My only concern is that `https://localhost:3001/#/jobs/undefined/1/output` is an example URL outputted from sparklines. This is not local to the details page though so I consider it non-blocking. Is this expected as this hasn't yet been implemented? If so, then this can go through just fine. Great work
